csrc/python_frontend/python_bindings_extension.cpp

```diff
@@ -10,5 +10,9 @@
 
 PYBIND11_MODULE(EXTENSION_NAME, m) {
   m.doc() = "nvfuser C API python binding"; // optional module docstring
+
   nvfuser::python_frontend::initNvFuserPythonBindings(m.ptr());
+
+  auto cleanup = []() -> void { nvfuser::python_frontend::cleanup(); };
+  m.add_object("_cleanup", py::capsule(cleanup));
```

Comment:
Please bear with my ignorance.

https://pybind11.readthedocs.io/en/stable/advanced/misc.html#module-destructors
vaguely remember the issue we had with clean up is about the nondeterministic destruction order here, so this reads like our selection to have communicator clean up to be called early is a conscious choice? Can we have a comment on that if so?

Reply:
My hope was that Communicator::cleanup is called **before** the cuda driver shutdown, which I believe happens during `atexit`. Therefore, in C++ unit tests, I'm calling it in testing::Environment::TearDown which happens before the end of main and therefore before `atexit`. 

I'm unsure when the module destructor is called (before `main` ends or after). It didn't trigger the crash at least when I ran `mpi_test.py`, so I'm guessing before. 

If your concern was about

> This approach has the potential downside that instances of classes exposed within the module may still be alive when the cleanup callback is invoked

I don't know enough to answer that... Yes, it looks scary but I'm not aware of a better option. Can you advise?

Reply:
> I don't know enough to answer that...

That makes the two of us. I thought there's a reason behind our choice between the two alternative and would like to have that in comments. I think among the alternatives listed in that doc, the one we went with is the better option since it has an earlier trigger.

I don't think further action is required at this point.